<a href="https://colab.research.google.com/github/anirbanghoshsbi/.github.io/blob/master/work/temp_hosting/Untitled114.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [38]:
import pandas as pd
import yfinance as yf
from sklearn.cluster import KMeans
from sklearn.preprocessing import StandardScaler


In [39]:

# Download data using yfinance
def get_data():
    data = pd.read_csv('/content/growth_oriented_companies.csv')
    data = data[~data['NSE Code'].str.contains('MCX')]
    data = data[~data['NSE Code'].str.contains('KSB')]
    data = data[~data['NSE Code'].str.contains('BHARATFORG')]

    return data

data=get_data()

In [40]:
data.shape

(166, 23)

In [41]:
data = data.drop(data.columns[0], axis=1)

In [42]:
data.columns

Index(['Name', 'BSE Code', 'NSE Code', 'Industry', 'Current Price',
       'Return on equity', 'Price to Earning', 'Price to Sales',
       'Debt to equity', 'Market Capitalization', 'Price to book value',
       'Return on capital employed', 'OPM', 'Current ratio',
       'Operating profit', 'Sales growth', 'Asset Turnover Ratio',
       'asset turnover ratio', 'Price to Free Cash Flow', 'Promoter holding',
       'Change in promoter holding', 'Cluster'],
      dtype='object')

In [43]:
features = [col for col in data.columns.tolist() if col not in ("BSE Code", "NSE Code",'Name','Industry','Current Price','Market Capitalization','Operating profit','asset turnover ratio', 'Price to Free Cash Flow', 'Promoter holding',
       'Change in promoter holding')]


In [44]:
features

['Return on equity',
 'Price to Earning',
 'Price to Sales',
 'Debt to equity',
 'Price to book value',
 'Return on capital employed',
 'OPM',
 'Current ratio',
 'Sales growth',
 'Asset Turnover Ratio',
 'Cluster']

In [45]:
scaled_data=data[features].fillna(0.001).values

In [46]:
# Extract the 'Cumulative Returns' column for clustering
X = scaled_data

# Perform KMeans clustering with 4 clusters
kmeans = KMeans(n_clusters=4, random_state=42)
data['Cluster'] = kmeans.fit_predict(X)

/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


In [47]:
# Analyze and interpret clusters
for i in range(4):
    cluster_data = data[data['Cluster'] == i]
    print(f"Cluster {i}:\n", cluster_data[features].median())
    print(f"Cluster_shape {i}:\n", cluster_data.shape)


Cluster 0:
 Return on equity              80.63
Price to Earning              60.03
Price to Sales                 9.25
Debt to equity                 0.04
Price to book value           42.07
Return on capital employed    79.13
OPM                           21.20
Current ratio                  1.42
Sales growth                   7.04
Asset Turnover Ratio           2.26
Cluster                        0.00
dtype: float64
Cluster_shape 0:
 (7, 22)
Cluster 1:
 Return on equity              18.250
Price to Earning              60.315
Price to Sales                 6.810
Debt to equity                 0.160
Price to book value            9.870
Return on capital employed    20.125
OPM                           18.970
Current ratio                  1.925
Sales growth                  13.710
Asset Turnover Ratio           1.015
Cluster                        1.000
dtype: float64
Cluster_shape 1:
 (116, 22)
Cluster 2:
 Return on equity               15.360
Price to Earning              102.655
P

In [48]:
data[data['Cluster']==2]

Name  BSE Code    NSE Code  \
1               A B B  500002.0         ABB   
4    Adani Energy Sol  539254.0  ADANIENSOL   
5       Adani Enterp.  512599.0    ADANIENT   
12   Apollo Hospitals  508869.0  APOLLOHOSP   
15             Astral  532830.0      ASTRAL   
16      Avenue Super.  540376.0       DMART   
32   Campus Activewe.  543523.0      CAMPUS   
36   Century Textiles  500040.0  CENTURYTEX   
47      Devyani Intl.  543330.0     DEVYANI   
49   Dixon Technolog.  540699.0       DIXON   
50                DLF  532868.0         DLF   
54            F A C T  590024.0        FACT   
61    Godrej Propert.  533150.0  GODREJPROP   
66   HDFC Life Insur.  540777.0    HDFCLIFE   
68   Hindustan Copper  513599.0  HINDCOPPER   
72     ICICI Pru Life  540133.0  ICICIPRULI   
78           JBM Auto  532605.0        JBMA   
79     Jubilant Food.  533155.0    JUBLFOOD   
82        Kaynes Tech  543664.0      KAYNES   
88        Laurus Labs  540222.0  LAURUSLABS   
92        Linde India  523457.0  LINDEINDIA   
94   Mahindra Holiday  533088.0       MHRIL   
99     Medplus Health  543427.0     MEDPLUS   
100      Metro Brands  543426.0  METROBRAND   
102              MMTC  513377.0        MMTC   
109  Olectra Greentec  532439.0     OLECTRA   
115    Pidilite Inds.  500331.0  PIDILITIND   
119    Radico Khaitan  532497.0      RADICO   
121   Relaxo Footwear  530517.0      RELAXO   
128     Shoppers Stop  532638.0  SHOPERSTOP   
129           Siemens  500550.0     SIEMENS   
130             Sobha  532784.0       SOBHA   
131  Solar Industries  532725.0   SOLARINDS   
137     Suzlon Energy  532667.0      SUZLON   
142     Tata Consumer  500800.0  TATACONSUM   
144   Tata Inv.Corpn.  501301.0  TATAINVEST   
145        Tata Steel  500470.0   TATASTEEL   
148     Titan Company  500114.0       TITAN   
153  United Breweries  532478.0         UBL   
157   Varun Beverages  540180.0         VBL   
160            Voltas  500575.0      VOLTAS   
161     Westlife Food  505533.0    WESTLIFE   

                                    Industry  Current Price  Return on equity  \
1                         Electric Equipment        5616.55             22.94   
4                Power Generation And Supply        1045.85             11.65   
5                                    Trading        3132.20              9.63   
12                                Healthcare        5983.20             13.71   
15                         Plastics Products        2030.20             17.91   
16                                   Trading        3932.10             15.95   
32                Leather / Leather Products         232.25             23.81   
36                                     Paper        1418.30              4.65   
47                                    Hotels         154.10             33.86   
49                    Electronics - Consumer        6962.00             22.52   
50                              Construction         823.70              5.47   
54                               Fertilizers         652.10             62.46   
61                              Construction        2218.25              5.39   
66                     Finance & Investments         632.35              9.57   
68                Mining / Minerals / Metals         271.15             14.73   
72                     Finance & Investments         567.75              8.45   
78                          Auto Ancillaries        1892.00             12.90   
79                                    Hotels         447.55             18.11   
82                  Electronics - Components        2666.60             16.38   
88   Pharmaceuticals - Indian - Formulations         400.90             21.39   
92                                 Chemicals        6441.10             18.18   
94                                    Hotels         383.00             35.92   
99                                   Trading         665.55              3.26   
100               Leather / Leather Products       

Here's a breakdown of Cluster 0 based on the financial metrics provided:

* **High profitability:** ROE (Return on Equity) of 80.63% and ROCE (Return on Capital Employed) of 79.13% indicate exceptional profitability. This means the company is generating high returns on the investment made by shareholders and on its capital employed.
* **High valuation:** P/E (Price to Earning) of 60.03 and P/B (Price to Book Value) of 42.07 suggest the company's stock might be trading at a premium relative to its earnings and book value. Investors seem to be willing to pay more due to the company's strong profitability.
* **Low debt:** D/E (Debt to Equity) ratio of 0.04 indicates very low debt levels, suggesting a financially conservative company.
* **Decent efficiency:** OPM (Operating Profit Margin) of 21.20% reflects a good ability to convert revenue into operating profit. Asset Turnover Ratio of 2.26 implies the company is efficiently generating sales from its assets.  
* **Moderate sales growth:** Sales growth of 7.04% is not exceptionally high but still positive.

Overall, Cluster 0 seems to represent companies with exceptional profitability, low debt, and decent efficiency. However, their high valuation suggests their stock prices might already reflect these positive aspects.

Let me know if you'd like to explore a specific company or metric in more detail.

Sure! Here's a breakdown of Cluster 1 and Cluster 2 based on the financial metrics provided:

**Cluster 1**

Cluster 1 seems to have companies with:

* Average profitability (ROE of 18.25%) and return on capital employed (ROCE of 20.13%)
* Moderate valuation (P/E of 60.32 and P/B of 9.87) compared to Cluster 0. This suggests their stock prices might be more reasonably priced relative to their earnings and book value.  
* Low debt (D/E of 0.16)  
* Decent efficiency (OPM of 18.97% and Asset Turnover Ratio of 1.02)

**Cluster 2**

Cluster 2 appears to consist of companies with:

* Lower profitability (ROE of 15.36% and ROCE of 15.04%) compared to Cluster 0 and 1
* Higher valuation (P/E of 102.66 and P/B of 13.14)  This could be due to reasons like higher growth prospects or intangible assets not fully captured in traditional financial metrics.
* More debt (D/E of 0.38) compared to Cluster 0 and 1  
* Modest efficiency (OPM of 14.23% and Asset Turnover Ratio of 1.11)

**Key Differences**

* Profitability: Cluster 0 > Cluster 1 > Cluster 2
* Valuation: Cluster 0 > Cluster 1 < Cluster 2
* Debt: Cluster 0 < Cluster 1 < Cluster 2

**In conclusion**

Clusters 1 and 2 represent companies with varying degrees of profitability, valuation, and financial leverage. Cluster 1 seems to balance these metrics more moderately, while Cluster 0 exhibits exceptional profitability but potentially high valuations. Companies in Cluster 2 might be riskier due to lower profitability, higher valuation, and more debt.

It's important to consider these findings along with other factors, such as industry and company-specific news, before making investment decisions.


In [49]:
data[data['Cluster']==0]

Name  BSE Code   NSE Code  \
28    Britannia Inds.  500825.0  BRITANNIA   
37   CG Power & Indu.  500093.0    CGPOWER   
41   Colgate-Palmoliv  500830.0     COLPAL   
91     Life Insurance  543526.0       LICI   
108      Nestle India  500790.0  NESTLEIND   
111     P & G Hygiene  500459.0       PGHH   
141         Tata Comm  500483.0   TATACOMM   

                                  Industry  Current Price  Return on equity  \
28                 Food - Processing - MNC        4974.20             66.62   
37                      Electric Equipment         469.70             66.05   
41           Personal Care - Multinational        2728.55             61.15   
91                   Finance & Investments         925.15            129.81   
108                Food - Processing - MNC        2603.65            108.18   
111          Personal Care - Multinational       15754.10             80.63   
141  Telecommunications - Service Provider        1934.85            141.54   

     Price to Earning  Price to Sales  Debt to equity  Market Capitalization  \
28              55.54            7.16            0.98              119812.71   
37              83.49            9.25            0.01               71738.84   
41              58.18           13.39            0.04               74212.64   
91              14.51            0.73            0.00              585157.16   
108             83.60           13.13            0.11              251032.78   
111             67.66           12.76            0.00               51138.97   
141             60.03            2.78            6.06               55143.23   

     ...    OPM  Current ratio  Operating profit  Sales growth  \
28   ...  19.02           0.87           3180.90          5.66   
37   ...  14.31           1.70           1110.09         17.96   
41   ...  32.86           1.42           1820.54          7.04   
91   ...   5.49           3.03          44168.56          1.56   
108  ...  23.38           0.95           4470.99         13.19   
111  ...  23.93           1.51            958.79          1.92   
141  ...  21.20           0.60           4208.02         13.19   

     Asset Turnover Ratio  asset turnover ratio  Price to Free Cash Flow  \
28                   2.37                  2.37                    83.94   
37                   1.69                  1.69                   210.75   
41                   2.12                  2.12                    65.30   
91                   4.33                  4.33                    15.88   
108                  2.77                  2.77                   131.45   
111                  2.26                  2.26                    71.90   
141                  1.24                  1.24                    22.72   

     Promoter holding  Change in promoter holding  Cluster  
28              50.55                        0.00        0  
37              58.11                       -0.01        0  
41              51.00                        0.00        0  
91              96.50                        0.00        0  
108             62.76                        0.00        0  
111             70.64                        0.00        0  
141             58.87                        0.01        0  

[7 rows x 22 columns]

In [50]:
data[data['Cluster']==1]

Name  BSE Code    NSE Code  \
0            3M India  523395.0     3MINDIA   
2    Aarti Industries  524208.0    AARTIIND   
3        Abbott India  500488.0  ABBOTINDIA   
6    Aether Industri.  543534.0      AETHER   
7         Affle India  542752.0       AFFLE   
..                ...       ...         ...   
159  Vijaya Diagnost.  543350.0      VIJAYA   
162   Whirlpool India  500238.0   WHIRLPOOL   
163          Yes Bank  532648.0     YESBANK   
164  Zee Entertainmen  505537.0        ZEEL   
165     ZF Commercial  533023.0   ZFCVINDIA   

                                      Industry  Current Price  \
0                          Diversified - Large       29594.45   
2                                    Chemicals         634.90   
3              Pharmaceuticals - Multinational       28404.35   
6                                    Chemicals         806.15   
7        Computers - Software - Medium / Small        1022.70   
..                                         ...            ...   
159                                 Healthcare         605.20   
162                        Domestic Appliances        1229.40   
163                     Banks - Private Sector          23.80   
164  Entertainment / Electronic Media Software         141.70   
165                           Auto Ancillaries       15505.35   

     Return on equity  Price to Earning  Price to Sales  Debt to equity  \
0               18.75             66.93            8.59            0.01   
2               11.55             53.03            3.68            0.64   
3               31.61             52.69           10.49            0.03   
6               15.91             82.58           16.08            0.01   
7               18.51             52.69            8.47            0.14   
..                ...               ...             ...             ...   
159             16.57             54.16           12.06            0.42   
162              6.37             77.17            2.30            0.01   
163              1.99             66.85            2.60            7.42   
164              0.35             62.52            1.59            0.02   
165             13.86             72.60            7.73            0.02   

     Market Capitalization  ...    OPM  Current ratio  Operating profit  \
0                 33352.95  ...  17.59           2.19            682.68   
2                 23015.38  ...  15.11           1.06            945.00   
3                 60359.24  ...  24.40           2.37           1403.69   
6                 10685.54  ...  27.34           6.77            181.66   
7                 14339.18  ...  19.56           1.91            331.02   
..                     ...  ...    ...            ...               ...   
159                6193.96  ...  40.28           4.05            206.88   
162               15597.40  ...   5.39           2.52            364.86   
163               68463.30  ...  56.62           3.30          14929.32   
164               13610.56  ...   9.94           4.29            852.53   
165               29398.14  ...  14.72           4.48            560.23   

     Sales growth  Asset Turnover Ratio  asset turnover ratio  \
0            7.91                  1.64                  1.64   
2           -6.90                  0.78                  0.78   
3            9.37                  1.81                  1.81   
6            8.07                  0.41                  0.41   
7           21.48                  0.73                  0.73   
..            ...                   ...                   ...   
159         13.02                  0.58                  0.58   
162          0.99                  1.26                  1.26   
163         23.03                  0.40                  0.40   
164          3.39                  0.68                  0.68   
165         17.12                  1.29                  1.29   

     Price to Free Cash Flow  Promoter holding  Change in promoter holding  \
0         

In [51]:
data[data['Cluster']==2]

Name  BSE Code    NSE Code  \
1               A B B  500002.0         ABB   
4    Adani Energy Sol  539254.0  ADANIENSOL   
5       Adani Enterp.  512599.0    ADANIENT   
12   Apollo Hospitals  508869.0  APOLLOHOSP   
15             Astral  532830.0      ASTRAL   
16      Avenue Super.  540376.0       DMART   
32   Campus Activewe.  543523.0      CAMPUS   
36   Century Textiles  500040.0  CENTURYTEX   
47      Devyani Intl.  543330.0     DEVYANI   
49   Dixon Technolog.  540699.0       DIXON   
50                DLF  532868.0         DLF   
54            F A C T  590024.0        FACT   
61    Godrej Propert.  533150.0  GODREJPROP   
66   HDFC Life Insur.  540777.0    HDFCLIFE   
68   Hindustan Copper  513599.0  HINDCOPPER   
72     ICICI Pru Life  540133.0  ICICIPRULI   
78           JBM Auto  532605.0        JBMA   
79     Jubilant Food.  533155.0    JUBLFOOD   
82        Kaynes Tech  543664.0      KAYNES   
88        Laurus Labs  540222.0  LAURUSLABS   
92        Linde India  523457.0  LINDEINDIA   
94   Mahindra Holiday  533088.0       MHRIL   
99     Medplus Health  543427.0     MEDPLUS   
100      Metro Brands  543426.0  METROBRAND   
102              MMTC  513377.0        MMTC   
109  Olectra Greentec  532439.0     OLECTRA   
115    Pidilite Inds.  500331.0  PIDILITIND   
119    Radico Khaitan  532497.0      RADICO   
121   Relaxo Footwear  530517.0      RELAXO   
128     Shoppers Stop  532638.0  SHOPERSTOP   
129           Siemens  500550.0     SIEMENS   
130             Sobha  532784.0       SOBHA   
131  Solar Industries  532725.0   SOLARINDS   
137     Suzlon Energy  532667.0      SUZLON   
142     Tata Consumer  500800.0  TATACONSUM   
144   Tata Inv.Corpn.  501301.0  TATAINVEST   
145        Tata Steel  500470.0   TATASTEEL   
148     Titan Company  500114.0       TITAN   
153  United Breweries  532478.0         UBL   
157   Varun Beverages  540180.0         VBL   
160            Voltas  500575.0      VOLTAS   
161     Westlife Food  505533.0    WESTLIFE   

                                    Industry  Current Price  Return on equity  \
1                         Electric Equipment        5616.55             22.94   
4                Power Generation And Supply        1045.85             11.65   
5                                    Trading        3132.20              9.63   
12                                Healthcare        5983.20             13.71   
15                         Plastics Products        2030.20             17.91   
16                                   Trading        3932.10             15.95   
32                Leather / Leather Products         232.25             23.81   
36                                     Paper        1418.30              4.65   
47                                    Hotels         154.10             33.86   
49                    Electronics - Consumer        6962.00             22.52   
50                              Construction         823.70              5.47   
54                               Fertilizers         652.10             62.46   
61                              Construction        2218.25              5.39   
66                     Finance & Investments         632.35              9.57   
68                Mining / Minerals / Metals         271.15             14.73   
72                     Finance & Investments         567.75              8.45   
78                          Auto Ancillaries        1892.00             12.90   
79                                    Hotels         447.55             18.11   
82                  Electronics - Components        2666.60             16.38   
88   Pharmaceuticals - Indian - Formulations         400.90             21.39   
92                                 Chemicals        6441.10             18.18   
94                                    Hotels         383.00             35.92   
99                                   Trading         665.55              3.26   
100               Leather / Leather Products       